### ISSUE TITLE
Turbo mode improvements

### ISSUE THREAD

**Issue by johnnovak** [dosbox-staging/dosbox-staging#2184]
I propose the following improvement options to be implemented for the "Alt-F12 turbo mode" because the current behaviour annoys the hell out of me (that's as good a user story as any! 🤣)

Specifically, I like to be able to toggle turbo mode on or off ("toggle" behaviour) as opposed to having to hold the hotkey down (current "momentary" behaviour). Other emulators such as VICE and WinUAE only provide "toggle" behaviour.

``` ini
turbo_mode = toggle | momentary
```

The other point of annoyance, "press-play-and-fast-forward-at-the-same-time-on-a-tape-deck" type of sound output while turbo is engaged is a nice gimmick (or is it?) but it's just unwanted screeching noise I don't care about (people born before 2000, you have no idea what I'm talking about, right? 😝) So I'd very much prefer to have at least an option to turn it off. Btw, sound is turned off in VICE and WinUAE in "warp mode", and there's no option to enable it.

``` ini
turbo_sound = on | off
```

_Important note: MIDI notes should be still generated & sent to FluidSynth and MT-32 in turbo mode; only the master sound output should be muted._

I think these options should be added to the `[dosbox]` section rather than the `[cpu]` section. Reasoning: setting cycles to max only affects CPU speed, whereas turbo mode creates a convenient little time-bubble around the emulated machine and accelerates time for the whole computer 😎 

Personally, I'd be happy to change the default behaviour to this:

``` ini
[dosbox]
turbo_mode = toggle
turbo_sound = off
```

### External MIDI

An interesting case is what would happen with external MIDI? Well, I assume a lot of interesting things, and I think the current behaviour of sending MIDI data at increased rates could result in funny and/or broken behaviour with some external sound modules that expect MIDI events at "normal real-time" rates (I haven't tested this yet, btw).

We can't just turn MIDI output off during turbo mode because the MIDI stream could contain controller change, bank/patch changes, and any random sysex data that would reprogram the MIDI device. So to do thins properly, we'd need to buffer all such controller & sysex data, and send them in one go once turbo mode is disengaged.

Btw, this is very similar to "MIDI event chasing", a feature implemented in all good MIDI sequencers that do a very similar thing when jumping to a different play position manually (ideally, the sequencer needs to replay the whole MIDI stream from the very beginning after each jump, "chasing" the controller events and replaying them before playback is commenced from the new playback position).

**Comment by kcgen** [comment]
Sounds great; I especially like your suggested defaults. 

We've got ToggleMute() in the mixer, which we use during the muted-turbo events. 

It in turn calls MIDI_Disengage(), which std::swap's the active midi device with a placeholder device, and then back again on _Engage() when unmuted.

This has worked well mostly because the block of sysex messages are sent upfront, followed by only channel changes. So on re-engagement, the midi device resumes playing the current channel messages. 

But I suppose if a sysex block was sent when muted, it would lose this. We could have the swapped in device queue sysex messages and then replay them on re-engagement.

**Comment by johnnovak** [comment]
> Sounds great; I especially like your suggested defaults.
> 
> We've got ToggleMute() in the mixer, which we could/should use during the muted-turbo events.
> 
> It in turn calls MIDI_Disengage(), which std::swap's the active midi device with a placeholder device, and then back again on _Engage() when unmuted.
> 
> This has worked well mostly because the block of sysex messages are sent upfront, followed by only channel changes. So on re-engagement, the midi device resumes playing the current channel messages.
> 
> But I suppose if a sysex block was sent when muted, it would lose this. We could have the swapped in device queue sysex messages and then replay them on re-engagement.

Yeah, I was writing what I wrote starting from first princinples and my experience with MIDI as a musician, but now that you mentioned it, I remember reviewing that MIDI muting related PR of yours @kcgen. I agree we should start with using what we already have, and only complicate it further if really necessary. You're correct in saying that most games just set up the sound device via SysEx messages at init and that's it, so that approach will work well for most (and potentially all) games.

**Comment by kcgen** [comment]
Was thinking a bit more about how we can mute MIDI, and maybe we can get away with just ignoring channel message during the mute phase. The benefit would be that sysex messages still hit the device.

In the current midi PR, I had some temporary logging where I monitored the timing of sysex messages, and they can be sent very rapidly: hundreds of them, back to back, all within a couple milliseconds  I think the games just blast them out in a loop, and let the MPU receiver deal with them. Based on this, I /think/ we could get away with just letting them flow out, even in fast forward mode, without having to queue and pace them (at least, we can try initially, and only add queuing if needed).

**Comment by johnnovak** [comment]
> I think the games just blast them out in a loop, and let the MPU receiver deal with them.

Yeah, it depends whether our emulated MPU would be also sped up during turbo mode, or not. If not, I guess it could work, but we need to try.

Btw, that's more or less compatible with my buffering idea; the MPU essentially would act as a buffer (if it works well enough).

**Comment by rderooy** [comment]
You may also consider an auto-off when the user hits a key or presses a mouse button. This is useful for getting through lengthy installers, but the moment user input is required, you probably don't want turbo any longer.

**Comment by rderooy** [comment]
I also consider "turbo" to be a rather poor name for this function. People familiar with DOS PC's will think this is akin to the Turbo button of old. But it is instead a fast-forward (or 'warp' if you want), function.

**Comment by johnnovak** [comment]
> You may also consider an auto-off when the user hits a key or presses a mouse button.

Good idea, we might play around with it and see how it works in practice.

> I also consider "turbo" to be a rather poor name for this function. People familiar with DOS PC's will think this is akin to the Turbo button of old. But it is instead a fast-forward (or 'warp' if you want), function.

I agree and it always gives me a pause as well because it reminds me of the turbo button on my 486 DX2/66. I'm happy to rename it to "warp mode" if no objections; that's a much better name with "history" attached to it.

**Comment by rderooy** [comment]
DOSBox-X implemented an auto-off by default (controllable via a config file setting), as leaving it on caused spurious events when the user pressed a key. e.g., pressing Enter would cause several Enter events to get registered by the DOS program.
```ini
[cpu]
#                                           turbo: Enables Turbo (Fast Forward) mode to speed up operations.
#                               stop turbo on key: If set, the Turbo mode will be automatically stopped if a keyboard input is detected.
#                         stop turbo after second: If a positive integer is specified, the Turbo function will last for specific seconds.
turbo                                           = false
stop turbo on key                               = true
stop turbo after second                         = 0
```

I'm also fine with a rename.

**Comment by johnnovak** [comment]
> DOSBox-X implemented an auto-off by default (controllable via a config file setting), as leaving it on caused spurious events when the user pressed a key. e.g., pressing Enter would cause several Enter events to get registered by the DOS program.

Yeah, and while the "stop turbo on key" feature could come in handy, ultimately I think they've chose the wrong workaround.

Both in VICE and WinUAE typing works perfectly fine in ~1000% warp mode; they clearly put in some extra logic to handle input in warp mode correctly. Same goes for the mouse and even double clicking. So regardless of whether we implement "stop turbo on key" thing, I think applying a similar special input handling when warp mode is on would be nice.

**Comment by GranMinigun** [comment]
> I'm happy to rename it to "warp mode" if no objections

Objection! Fast-forward, if you are going for a rename. No need to reinvent the wheel or spawn even more confusion ("warp" isn't clear to me, for instance). There are also "turbo buttons" originating from old Famiclone controllers.

**Comment by kcgen** [comment]
Fast forward make sense to me, too.

If anyone wants to work on this - the "rate of fast-forwarding" can be throttled by shrinking the delay value inside the 1ms tick-loop.

This is in `src/dosbox.cpp`: 

```c++
constexpr auto duration = std::chrono::milliseconds(1);
std::this_thread::sleep_for(duration);
```

To fast-forward at 1.25x, use `duration = std::chrono::microseconds(750);`

Or 2x: use `duration = std::chrono::microseconds(500);`

It works in reverse too, slow-motion at 50% speed, use `std::chrono::milliseconds(2);` 

Fun stuff!

**Comment by Jkapp76** [comment]
I like all of this.
I like the FFWD audio ON, I use the sound as a cue on a game or two so I know when to let off. But if the audio could be improved while warping that could be better.

I like X's "stop turbo on key" too. It's nice to slap any key to stop it... sometimes. I like to skip long adventure game intros, like (I have no mouth and must scream.) using this feature. Also Space Quest 4, 5, and 6.
...I even used X's (stop turbo after x seconds) to auto-zip past the scream intro every time and land right at the game.

**Comment by johnnovak** [comment]
I like the "fast foward" suggestion; let's call it that. Thanks @GranMinigun 🎉 

"stop turbo on key" — I don't like this one, it's a redundant feature. To achieve this, you could just set `turbo_mode = toggle ` and toggle fast-forward off with the hotkey when you want to go back to real-time speed.

I also think it's an ill-conceived feature in DOSBox-X; my understanding is that it was added as a stopgap measure to prevent key-repeat going haywire in "turbo mode". A better implementation in other emulators (WinUAE, VICE) is to handle all input correctly _even_ in "turbo mode" with some trickery. In WinUAE and VICE you get normal key-repeat _even_ at 1000%+ emulation speeds! Even double-clicking with the mouse works 100% correctly in WinUAE at max fast forward speed! I think that's the ideal we should strive for instead of starting out with workarounds.

> I like the FFWD audio ON, I use the sound as a cue on a game or two so I know when to let off. But if the audio could be improved while warping that could be better.

That's fine, you could just set `turbo_sound = on` in your config then.

**Comment by Ohrwerk** [comment]
I would very much celebrate a turbo function.
I always use "pictview" to display the box covers of the games before they start.
With old games for which it is necessary to configure fixed cycles, it takes a long time to load the images.
I would really appreciate a config option like this.
CONFIG -set turbo=false/true
